File: tests/test_server_hardening.py
import pytest
import requests
import json
import threading
import time
from zena_mode.server import ThreadingHTTPServer, ZenAIOrchestrator
from config_system import config


def test_request_size_limit():
    """Verify that the 10MB request size limit is enforced."""
    # We'll use a local instance for testing to avoid messing with the main one
    test_port = 9010

    def run_server():
        """Run server."""
        try:
            server = ThreadingHTTPServer(("127.0.0.1", test_port), ZenAIOrchestrator)
            server.serve_forever()
        except Exception:
            pass

    t = threading.Thread(target=run_server, daemon=True)
    t.start()
    time.sleep(1)  # Wait for startup

    # Payload > 10MB (Assuming config.MAX_FILE_SIZE is 10 * 1024 * 1024)
    # Let's check the config value first in the test
    limit = config.MAX_FILE_SIZE
    large_data = "x" * (limit + 1024)

    try:
        url = f"http://127.0.0.1:{test_port}/api/chat"
        resp = requests.post(url, data=json.dumps({"message": large_data}, timeout=30), timeout=5)
        assert resp.status_code == 413, f"Expected 413 for large payload, got {resp.status_code}"
    except (requests.exceptions.ConnectionError, requests.exceptions.ChunkedEncodingError):
        # This is also a success: the server rejected the large payload by dropping the connection
        pass
    finally:
        pass  # In a real test we'd kill the server, but for now we let it die with the thread


def test_local_only_bind():
    """Verify that the server rejects non-local binds if possible to test."""
    # This is hard to test without multiple IPs, but we can verify the code-level change.
    # For now, we'll verify that it DOES work on 127.0.0.1
    url = f"http://127.0.0.1:{config.mgmt_port}/startup/progress"
    try:
        resp = requests.get(url, timeout=1)
        # If it responds at all, it's alive. status_code 200 or 404 is fine as long as connection is made.
        assert resp.status_code in [200, 404]
    except requests.exceptions.ConnectionError:
        # If the server isn't running, this is expected in some environments,
        # but for a 'tough' test we usually expect the server to be up or start it.
        pytest.skip("Server not running on mgmt_port")


def test_invalid_json_handling():
    """Verify that malformed JSON doesn't crash the orchestrator."""
    test_port = 9011

    def run_server():
        """Run server."""
        try:
            server = ThreadingHTTPServer(("127.0.0.1", test_port), ZenAIOrchestrator)
            server.serve_forever()
        except Exception:
            pass

    t = threading.Thread(target=run_server, daemon=True)
    t.start()
    time.sleep(1)

    url = f"http://127.0.0.1:{test_port}/api/chat"
    resp = requests.post(url, data="{invalid json...", headers={"Content-Type": "application/json"}, timeout=30)
    # BaseZenHandler.parse_json_body returns {} on failure, and handler returns 404 because path check might fail or return {}
    # But it shouldn't be a 500 CRASH.
    assert resp.status_code != 500

File: rust_output/src/test_server_hardening.rs
use anyhow::{Result, Context};
use crate::config_system::{config};
use crate::server::{ThreadingHTTPServer, ZenAIOrchestrator};
use std::collections::HashMap;

/// Verify that the 10MB request size limit is enforced.
pub fn test_request_size_limit() -> Result<()> {
    // Verify that the 10MB request size limit is enforced.
    let mut test_port = 9010;
    let run_server = || {
        // Run server.
        // try:
        {
            let mut server = ThreadingHTTPServer(("127.0.0.1".to_string(), test_port), ZenAIOrchestrator);
            server::serve_forever();
        }
        // except Exception as _e:
    };
    let mut t = std::thread::spawn(|| {});
    t.start();
    std::thread::sleep(std::time::Duration::from_secs_f64(1));
    let mut limit = config::MAX_FILE_SIZE;
    let mut large_data = ("x".to_string() * (limit + 1024));
    // try:
    {
        let mut url = format!("http://127.0.0.1:{}/api/chat", test_port);
        let mut resp = /* reqwest::post( */url, /* data= */ serde_json::to_string(&HashMap::from([("message".to_string(), large_data)])).unwrap(), /* timeout= */ 5);
        assert!(resp.status_code == 413, "Expected 413 for large payload, got {}", resp.status_code);
    }
    // except (requests.exceptions::ConnectionError, requests.exceptions::ChunkedEncodingError) as _e:
    // finally:
        // pass
}

/// Verify that the server rejects non-local binds if possible to test.
pub fn test_local_only_bind() -> Result<()> {
    // Verify that the server rejects non-local binds if possible to test.
    let mut url = format!("http://127.0.0.1:{}/startup/progress", config::mgmt_port);
    // try:
    {
        let mut resp = /* reqwest::get( */&url).cloned().unwrap_or(/* timeout= */ 1);
        assert!(vec![200, 404].contains(&resp.status_code));
    }
    // except requests.exceptions::ConnectionError as _e:
}

/// Verify that malformed JSON doesn't crash the orchestrator.
pub fn test_invalid_json_handling() -> Result<()> {
    // Verify that malformed JSON doesn't crash the orchestrator.
    let mut test_port = 9011;
    let run_server = || {
        // Run server.
        // try:
        {
            let mut server = ThreadingHTTPServer(("127.0.0.1".to_string(), test_port), ZenAIOrchestrator);
            server::serve_forever();
        }
        // except Exception as _e:
    };
    let mut t = std::thread::spawn(|| {});
    t.start();
    std::thread::sleep(std::time::Duration::from_secs_f64(1));
    let mut url = format!("http://127.0.0.1:{}/api/chat", test_port);
    let mut resp = /* reqwest::post( */url, /* data= */ "{invalid json...".to_string(), /* headers= */ HashMap::from([("Content-Type".to_string(), "application/json".to_string())]), /* timeout= */ 30);
    assert!(resp.status_code != 500);
}
